<a href="https://colab.research.google.com/github/prideven/DataMining/blob/main/DataCleaning_Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**`Data Cleaning`**

Performed data cleaning to predict the cancellation by the guest. 

Import Useful Functions and libraries

In [ ]:
# importing necessary libraries
import pandas as  pd 
import numpy as np 
import matplotlib.pyplot as plt
import csv
import gdown
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder
%matplotlib inline 
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [ ]:
# Loading dataset with pandas
#to load file from google drive
url='https://drive.google.com/uc?id=1G-v_gn4kyrPti3snkI-TovKyAc3TvaAB'
output= "data.csv"
gdown.download(url,output,quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1G-v_gn4kyrPti3snkI-TovKyAc3TvaAB
To: /content/data.csv
16.9MB [00:00, 218MB/s]


'data.csv'

In [ ]:
#loading the csv file into pandas dataframe
df=pd.read_csv("data.csv")

In [ ]:
# Display dataframe
df

hotel  is_canceled  ...  reservation_status  reservation_status_date
0       Resort Hotel            0  ...           Check-Out               2015-07-01
1       Resort Hotel            0  ...           Check-Out               2015-07-01
2       Resort Hotel            0  ...           Check-Out               2015-07-02
3       Resort Hotel            0  ...           Check-Out               2015-07-02
4       Resort Hotel            0  ...           Check-Out               2015-07-03
...              ...          ...  ...                 ...                      ...
119385    City Hotel            0  ...           Check-Out               2017-09-06
119386    City Hotel            0  ...           Check-Out               2017-09-07
119387    City Hotel            0  ...           Check-Out               2017-09-07
119388    City Hotel            0  ...           Check-Out               2017-09-07
119389    City Hotel            0  ...           Check-Out               2017-09-07

[119390 rows x 32 columns]

In [ ]:
#display only few elements
df.head

<bound method NDFrame.head of                hotel  is_canceled  ...  reservation_status  reservation_status_date
0       Resort Hotel            0  ...           Check-Out               2015-07-01
1       Resort Hotel            0  ...           Check-Out               2015-07-01
2       Resort Hotel            0  ...           Check-Out               2015-07-02
3       Resort Hotel            0  ...           Check-Out               2015-07-02
4       Resort Hotel            0  ...           Check-Out               2015-07-03
...              ...          ...  ...                 ...                      ...
119385    City Hotel            0  ...           Check-Out               2017-09-06
119386    City Hotel            0  ...           Check-Out               2017-09-07
119387    City Hotel            0  ...           Check-Out               2017-09-07
119388    City Hotel            0  ...           Check-Out               2017-09-07
119389    City Hotel            0  ...        

In [ ]:
#to display the total number of rows and columns present in the data
df.shape

(119390, 32)

The dimension of our dataset is 119390 rows and 32 columns.
We can now perform cleaning activity to extract some information like the number of non-null values in every column and their datatypes

In [ ]:
#to display the valid information present in the dataset
df.info

<bound method DataFrame.info of                hotel  is_canceled  ...  reservation_status  reservation_status_date
0       Resort Hotel            0  ...           Check-Out               2015-07-01
1       Resort Hotel            0  ...           Check-Out               2015-07-01
2       Resort Hotel            0  ...           Check-Out               2015-07-02
3       Resort Hotel            0  ...           Check-Out               2015-07-02
4       Resort Hotel            0  ...           Check-Out               2015-07-03
...              ...          ...  ...                 ...                      ...
119385    City Hotel            0  ...           Check-Out               2017-09-06
119386    City Hotel            0  ...           Check-Out               2017-09-07
119387    City Hotel            0  ...           Check-Out               2017-09-07
119388    City Hotel            0  ...           Check-Out               2017-09-07
119389    City Hotel            0  ...      

In [ ]:
# find the total count of null values in each columns. 
np.sum(df.isnull())

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

We can see from above dataset, it contains 7 columns with null values. The null values need to be either filled with a default values or it can omitted with respect to column elimination or row elimination. 

**REMOVING UNWANTED DATA**

Usually, if more than 70% of values in a column are missing and there is no way to fill in the missing values, then the column can be dropped completely from the dataset.

**i)REMOVE Null VALUES**

Usually, if more than 70% of values in a column are missing and there is no way to fill in the missing values, then the column can be dropped completely from the dataset.

In [ ]:
for column in df.columns:
  if np.sum(df[column].isnull())>(df.shape[0]*0.7):
    df.drop(columns=column,inplace=True,axis=1)
print(df.shape)

(119390, 31)


The company column which contained more than 1 lakh missing values is removed. Hence, the total count of columns is 31 at this point.

**ii)REMOVE Redudant data**

The “arrival_date_week_number” column is of no use as we already have 3 columns of the year, month, and the day giving us the date of arrival. Also as we have created a new column showing the date, we no more need the 3 individual columns. Hence let us remove these column

In [ ]:
df=df.drop(columns=["arrival_date_week_number","arrival_date_day_of_month","arrival_date_year","arrival_date_month"])



In [ ]:
print(df.shape)

(119390, 27)


After removing 4 more columns, the total count of columns is 28 at this point.

**iii) Removing unwanted rows**

Depending upon what values we are predicting, we can either remove the entire column or remove the row having empty values.
Now below step shows to remove all rows having a missing value in the agent col

In [ ]:
df.dropna(subset=["agent"],inplace=True)
df.shape

(103050, 27)

We can see the number of rows has consideralbly decreased, using the dropna function on our dataframe. 

**Filling the missing values of columns**

As mean can be of float datatype,we can take the lowest nearest integer using the floor method

In [ ]:
df['children'].fillna(value=df["children"].mean(),inplace=True)
df["children"]=df["children"].apply(np.floor)
totalNull=np.sum(df.children.isnull())
print(totalNull)
dataframe=df

0


In [ ]:
df.head

<bound method NDFrame.head of                hotel  is_canceled  ...  reservation_status  reservation_status_date
3       Resort Hotel            0  ...           Check-Out               2015-07-02
4       Resort Hotel            0  ...           Check-Out               2015-07-03
5       Resort Hotel            0  ...           Check-Out               2015-07-03
7       Resort Hotel            0  ...           Check-Out               2015-07-03
8       Resort Hotel            1  ...            Canceled               2015-05-06
...              ...          ...  ...                 ...                      ...
119385    City Hotel            0  ...           Check-Out               2017-09-06
119386    City Hotel            0  ...           Check-Out               2017-09-07
119387    City Hotel            0  ...           Check-Out               2017-09-07
119388    City Hotel            0  ...           Check-Out               2017-09-07
119389    City Hotel            0  ...        

**Dropping off less related data** 

In [ ]:
df_subset = df.drop([
                           "hotel","meal", 'assigned_room_type','reserved_room_type','reservation_status_date',
                            'previous_cancellations','previous_bookings_not_canceled','reservation_status','customer_type','deposit_type','distribution_channel',"market_segment","country"],axis=1)

print the values after dropping the data



In [ ]:
df_subset

is_canceled  ...  total_of_special_requests
3                 0  ...                          0
4                 0  ...                          1
5                 0  ...                          1
7                 0  ...                          1
8                 1  ...                          1
...             ...  ...                        ...
119385            0  ...                          0
119386            0  ...                          2
119387            0  ...                          4
119388            0  ...                          0
119389            0  ...                          2

[103050 rows x 14 columns]

**Train Test Split**


Splitting the dataset into train and test. The default size of the split ratio is 3:1 is performed. 

In [ ]:
def data_split(df, label):
    
    from sklearn.model_selection import train_test_split

    X = df.drop(label, axis=1)
    Y = df[label]

    x_train, x_test, y_train, y_test = train_test_split(X,Y,random_state=0)
    
    return x_train, x_test, y_train, y_test
x_train, x_test, y_train, y_test = data_split(df_subset, 'is_canceled')

**Model using Decision Tree**

There are many ways to predict our model and i have used decision tree to predict the model.


In [ ]:
def train(x_train, y_train):
    clf = DecisionTreeClassifier(random_state=0)
    clf.fit(x_train,y_train)
    return clf
clf = train(x_train, y_train)

**Evaluating the Machine learning Model**

To evaluate our model we need to check the training and testing accuracy

In [ ]:
def Score(clf,x_train,y_train,x_test,y_test):
    train_score = clf.score(x_train,y_train)
    test_score = clf.score(x_test,y_test)

    print("========================================")
    print(f'Training Accuracy of model: {train_score}')
    print(f'Test Accuracy of model is: {test_score}')
    print("========================================")
    
    
Score(clf,x_train,y_train,x_train,y_train)

Training Accuracy of model: 0.9882127654068601
Test Accuracy of model is: 0.9882127654068601


**Random value Prediction**

Picking a random sample  to make the prediction and compare it with the actual values

In [ ]:
## Getting Prediciton of 100th record of x_train
prediction = clf.predict(x_train.iloc[100].values.reshape(1,-1))

## Actual Value of 100th record of x_train from y_train
actual_value = y_train.iloc[100]

print(f'Predicted Value \t: {prediction[0]}')
print(f'Actual Value\t\t: {actual_value}')

Predicted Value 	: 0
Actual Value		: 0


Our model correctly predicted that the guest will not cancel the booking.

**COCLUSION:**

Made the predictive model using the Decision Tree to predict whether the guest will cancel the booking or not and achieved 98% accuracy.